lib/activerecord_autoreplica.rb

```diff
@@ -91,12 +91,17 @@ def self.using_read_replica_pool(replica_connection_pool)
   def self.in_replica_context(read_pool)
     return yield if current_read_pool # This method should not be reentrant
 
-    begin
-      self.current_read_pool = read_pool
-      yield
-    ensure
-      current_read_pool.release_connection
-      clear_current_read_pool
+    Thread.handle_interrupt(Exception => :never) do
```

Comment:
I'm not sure about why this is needed: this method was made specifically to be safe from timeouts, as it only does 1 atomic operation before yielding.

Reply:
Emphasis on "before". What this method tries to guarantee is the `ensure` block — iff. a connection is taken from the pool, it'll get released.
If there is an interrupt at anypoint after this atomic operation before the block,  the `ensure` block might never get run. Or get partially run.

So, both the "before" and "after" part have to be unnterruptible, unlike the yielded part, if that makes sense?

Reply:
For more info:
* https://jvns.ca/blog/2015/11/27/why-rubys-timeout-is-dangerous-and-thread-dot-raise-is-terrifying/
* http://blog.headius.com/2008/02/ruby-threadraise-threadkill-timeoutrb.html

That's exactly [what `Rack::Timeout` does](https://github.com/heroku/rack-timeout/blob/master/lib/rack/timeout/support/timeout.rb#L6).

Reply:
Yes I know! That's exactly while we created this fork :) The idea is that everything that happens here is safely interruptible:
- The set operation is atomic
- `current_read_pool.release_connection` _can_ fail, but the connection will be reaped by the connection pool (as it would if the thread died e.g. during the query execution) .
- `clear_current_read_pool` does not matter if the thread is terminated.

Reply:
> `current_read_pool.release_connection` can fail, but the connection will be reaped by the connection pool (as it would if the thread died e.g. during the query execution

where would it be reaped? by what code?
(it's very probable from errors we're seeing that the connections _do not_ get returned to the pool)

Reply:
My understanding is that this method https://apidock.com/rails/v3.2.13/ActiveRecord/ConnectionAdapters/ConnectionPool/clear_stale_cached_connections%21 will be invoked in Rails 3 if there is no available connection.

However if we are worried about the connections not being returned to the pool, I think with this PR it can still happen within the `yield`: it's much more likely that the timeout is going to hit there than within this wrapper.

Reply:
> My understanding is that this method `ConnectionPool#clear_stale_cached_connections` will be invoked in Rails 3 if there is no available connection.

Indeed (it's called on (ConnectionPool#checkout`). Not sure _why_ it wouldn't work properly; possibly because we're still in the same (live) thread and the connections aren't considered "stale".

There's also the the `ActiveRecord::ConnectionAdapters::ConnectionManagement` middleware. Unfortunately that only cleans up the pool managed by Rails - not the read pool that's instantiated manually here.

> However if we are worried about the connections not being returned to the pool, I think with this PR it can still happen within the yield: it's much more likely that the timeout is going to hit there than within this wrapper.

Timeouts can still happen, yes - it's fine if they happen in the `yield`, because them the error bubbles out of `yield` and the `ensure` block gets run. What we're protecting against is cases when what's _outside_ the yield (including the `ensure` block) gets interrupted.

Reply:
True, it would indeed be catched by ensure inside `yield`.

I guess it might well be that the `release_connection` itself leaves things in an undesirable state when interrupted; in this case, I think your solution should solve it 👍